util/utility.js

```diff
@@ -315,68 +319,47 @@ function getData(redis, url, cb) {
-    request({
-      url: target,
-      json: hypixelApi,
+  await pTimeout(delay);
+  try {
+    const { body } = await got(target, {
+      responseType: hypixelApi ? 'json' : 'text',
       timeout,
-    }, (err, res, body) => {
-      if (err
-        || !res
-        || res.statusCode !== 200
-        || !body
-        || (hypixelApi && !body.success)
-      ) {
-        // invalid response
-        if (url.noRetry) {
-          return cb('invalid response');
-        }
-        if (mojangApi) {
-          return cb('Failed to get player uuid', null);
-        }
-        if (body) {
-          logger.error(`[INVALID] error: ${body.cause}, retrying ${target}`);
-          if (body.cause === 'Internal error') {
-            return cb('Internal error');
-          }
-        } else {
-          logger.error(`[INVALID] status: ${res ? res.statusCode : ''}, retrying ${target}`);
-        }
-        return getData(redis, url, cb);
-      }
-      if (hypixelApi && !body.success) {
-        // valid response, but invalid data, retry
-        if (url.noRetry) {
-          return cb('invalid data');
-        }
-        // insert errors to redis for monitoring
-        let failed = 0;
-        const multi = redis.multi()
-          .incr('hypixel_api_error')
-          .expireat('hypixel_api_error', getStartOfBlockMinutes(1, 1));
-        multi.exec((err, resp) => {
-          if (err) {
-            logger.error(err);
-          }
-          [failed] = resp;
-          logger.warn(`Failed API requests in the past minute: ${failed}`);
-          logger.error(`[INVALID] data: ${target}, retrying ${JSON.stringify(body)}`);
-          let backoff = (body.throttle)
-            ? 3000
-            : 0;
-          backoff += (failed > 50)
-            ? 10 * failed
-            : 0;
-          logger.debug(`getData timout for ${backoff}ms`);
-          return setTimeout(() => {
-            getData(redis, url, cb);
-          }, backoff);
-        });
-      }
-      return cb(null, body);
+      retry: url.noRetry ? 0 : 99,
+      hooks: {
+        afterResponse: [
+          async (response, retryWithMergedOptions) => {
+            if (hypixelApi && !response.body.success) {
+              const multi = redis.multi()
+                .incr('hypixel_api_error')
+                .expireat('hypixel_api_error', getStartOfBlockMinutes(1, 1));
+
+              try {
+                const [failed] = await promisify(multi.exec);
+                logger.warn(`Failed API requests in the past minute: ${failed}`);
+                logger.error(`[INVALID] data: ${target}, retrying ${JSON.stringify(body)}`);
+              } catch (err) {
+                logger.error(err);
+              }
+              return retryWithMergedOptions();
+            }
+            return response;
+          },
+        ],
+      },
     });
-  }, delay);
-}
+
+    return body;
+  } catch (error) {
+    if (url.noRetry) {
+      throw new Error('Invalid response');
```

Comment:
How does throwing an error affect the data passed?

Reply:
Since this is an async function, universalify puts the error as the first argument in the callback function. The only change we need to make to calling functions is to take `err.message` instead of the `err` object directly.